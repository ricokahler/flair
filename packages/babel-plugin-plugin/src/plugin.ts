import fs from 'fs';
import _path from 'path';
import { Visitor } from '@babel/traverse';
import * as t from '@babel/types';
import {
  collect,
  transformCssTemplateLiteral,
  CollectionPluginOptions,
} from '@react-style-system/collect';
import { seek, createFilenameHash } from '@react-style-system/common';

interface Options extends CollectionPluginOptions {
  themePath: string;
  cacheDir: string;
}

const importSourceValue = 'react-style-system';
const replacementImportSourceValue = '@react-style-system/ssr';
const importedName = 'createStyles';

function plugin(
  _: any,
  opts: Options,
): {
  visitor: Visitor<{ opts: Options; file: { opts: { filename: string } } }>;
} {
  const { cacheDir } = opts;

  return {
    visitor: {
      Program(path, state) {
        /**
         * this is used to match the `useStyle` index. because we allow more than one
         * `createStyles` call, we need to keep track of what number this one is.
         */
        let useStyleIndex = 0;

        const { filename } = state.file.opts;
        const filenameHash = createFilenameHash(filename);
        const cssFilename = _path.join(cacheDir, `${filenameHash}.css`);

        // Find create styles before continuing
        const foundCreateStyles = seek<boolean>(report => {
          path.traverse({
            ImportDeclaration(path) {
              const { specifiers, source } = path.node;

              const hasPackageName = source.value === importSourceValue;
              if (!hasPackageName) return;

              const hasCreateStyles = specifiers.some(node => {
                if (!t.isImportSpecifier(node)) return false;
                return node.imported.name === importedName;
              });
              if (!hasCreateStyles) return;

              report(true);
            },
          });

          report(false);
        });
        if (!foundCreateStyles) return;

        // Generate the CSS via `collect`
        // This generation is done up here so that this plugin can early quit
        // if `collect` throws
        const css = collect(filename, opts);

        // Add the import for the CSS filename
        path.node.body.unshift(
          t.importDeclaration([], t.stringLiteral(cssFilename)),
        );

        path.traverse({
          // Rewrite imports
          ImportDeclaration(path) {
            const { specifiers, source } = path.node;
            const hasCreateStyles = specifiers.some(node => {
              if (!t.isImportSpecifier(node)) return false;
              return node.imported.name === importedName;
            });
            if (!hasCreateStyles) return;

            const hasPackageName = source.value === importSourceValue;
            if (!hasPackageName) return;

            path.node.source = t.stringLiteral(replacementImportSourceValue);
          },

          // Rewrite `createStyles` return object expression
          CallExpression(path) {
            const { callee, arguments: expressionArguments } = path.node;
            if (!t.isIdentifier(callee)) return;
            if (callee.name !== importedName) return;

            const [firstArgument] = expressionArguments;

            if (
              !t.isFunctionExpression(firstArgument) &&
              !t.isArrowFunctionExpression(firstArgument)
            ) {
              return;
            }

            const { body } = firstArgument;

            const stylesObjectExpression = seek<t.ObjectExpression>(report => {
              if (t.isObjectExpression(body)) {
                report(body);
              }

              path.traverse({
                ReturnStatement(path) {
                  const { argument } = path.node;

                  if (!t.isObjectExpression(argument)) return;
                  report(argument);
                },
              });
            });

            // Go through each property and replace it with strings that can be
            // replaced with CSS variables
            stylesObjectExpression.properties = stylesObjectExpression.properties.map(
              property => {
                if (!t.isObjectProperty(property)) return property;

                const { value, key } = property;
                if (!t.isTaggedTemplateExpression(value)) return property;

                const { tag, quasi } = value;
                if (!t.isIdentifier(tag)) return property;
                if (tag.name !== 'css') return property;

                const transformedTemplateLiteral = transformCssTemplateLiteral(
                  quasi,
                );

                const cssPropertyExpressions = transformedTemplateLiteral.expressions.filter(
                  expression => {
                    if (
                      t.isCallExpression(expression) &&
                      t.isIdentifier(expression.callee) &&
                      expression.callee.name === 'staticVar'
                    ) {
                      return false;
                    }
                    return true;
                  },
                );

                return t.objectProperty(
                  key,
                  t.arrayExpression(cssPropertyExpressions),
                );
              },
            );

            stylesObjectExpression.properties.push(
              t.objectProperty(
                t.identifier('classNamePrefix'),
                t.stringLiteral(
                  `${createFilenameHash(filename)}-${useStyleIndex}`,
                ),
              ),
            );

            useStyleIndex += 1;
          },
        });

        // Tradeoff alert:
        //
        // using fs.writeFileSync allows us to write a file from a babel plugin.
        // this means we don't need any bundler integration for this thing to
        // work but it's kind of weird to see this kind of side-effect inside of
        // a babel plugin
        if (!fs.existsSync(cacheDir)) {
          fs.mkdirSync(cacheDir);
        }
        fs.writeFileSync(cssFilename, css);
      },
    },
  };
}

export default plugin;
