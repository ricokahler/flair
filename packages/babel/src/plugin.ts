import fs from 'fs';
import _path from 'path';
import { Visitor } from '@babel/traverse';
import * as t from '@babel/types';
import collect from '@react-style-system/collect';
import createFileNameHash from './createFileNameHash';

interface Result {
  visitor: Visitor;
}

function range(n: number) {
  return Array.from(Array(n)).map((_, i) => i);
}

function plugin(state: any, opts: any): Result {
  function createArrayPropertyValueFromTemplateLiteral(
    quasi: t.TemplateLiteral,
  ) {
    const { expressions, quasis } = quasi;

    const cssPropertyExpressions = range(expressions.length)
      .map(i => ({
        expression: expressions[i],
        templateElement: quasis[i],
      }))
      .filter(({ templateElement }) => {
        // must end with a `:` to signify that the expression is a CSS property
        return templateElement.value.raw.trim().endsWith(':');
      })
      .map(({ expression }) => expression);

    return t.arrayExpression(cssPropertyExpressions);
  }

  const { cacheDir } = opts;

  return {
    visitor: {
      Program(path, state: any) {
        const { filename } = state.file.opts;
        const filenameHash = createFileNameHash(filename);
        const cssFilename = _path.join(cacheDir, `${filenameHash}.css`);

        let foundCreateStyles = false;
        path.traverse({
          ImportDeclaration(path) {
            const {
              importSourceValue = 'react-style-system',
              importedName = 'createStyles',
            } = state.opts;

            const { specifiers, source } = path.node;
            const hasCreateStyles = specifiers.some(node => {
              if (!t.isImportSpecifier(node)) return false;
              return node.imported.name === importedName;
            });
            if (!hasCreateStyles) return;

            const hasPackageName = source.value === importSourceValue;
            if (!hasPackageName) return;

            foundCreateStyles = true;
          },
        });

        if (!foundCreateStyles) return;

        path.node.body.unshift(
          t.importDeclaration([], t.stringLiteral(cssFilename)),
        );

        path.traverse({
          ImportDeclaration(path) {
            const {
              importSourceValue = 'react-style-system',
              importedName = 'createStyles',
            } = state.opts;

            const { specifiers, source } = path.node;
            const hasCreateStyles = specifiers.some(node => {
              if (!t.isImportSpecifier(node)) return false;
              return node.imported.name === importedName;
            });
            if (!hasCreateStyles) return;

            const hasPackageName = source.value === importSourceValue;
            if (!hasPackageName) return;

            path.node.source = t.stringLiteral(`${importSourceValue}/ssr`);

            const { filename } = state.file.opts;
            const css = collect(filename, opts);
            if (!css) return;
            const filenameHash = createFileNameHash(filename);
            const cssFilename = _path.join(cacheDir, `${filenameHash}.css`);
            fs.writeFileSync(cssFilename, css);
          },

          CallExpression(path) {
            const { opts } = state;
            const { filename } = state.file.opts;
            const { importedName = 'createStyles' } = opts;

            const { callee, arguments: expressionArguments } = path.node;
            if (!t.isIdentifier(callee)) return;
            if (callee.name !== importedName) return;

            const [firstArgument] = expressionArguments;

            if (
              !t.isFunctionExpression(firstArgument) &&
              !t.isArrowFunctionExpression(firstArgument)
            ) {
              return;
            }

            const { body } = firstArgument;

            let stylesObjectExpression!: t.ObjectExpression;
            if (t.isObjectExpression(body)) {
              stylesObjectExpression = body;
            } else {
              path.traverse({
                ReturnStatement(path) {
                  const { argument } = path.node;

                  if (!t.isObjectExpression(argument)) return;
                  stylesObjectExpression = argument;
                },
              });
            }

            for (const property of stylesObjectExpression.properties) {
              if (!t.isObjectProperty(property)) return;
              const { value } = property;

              if (!t.isTaggedTemplateExpression(value)) return;
              const { tag, quasi } = value;

              if (!t.isIdentifier(tag)) return;
              if (tag.name !== 'css') return;

              const arrayExpression = createArrayPropertyValueFromTemplateLiteral(
                quasi,
              );

              property.value = arrayExpression;
            }

            stylesObjectExpression.properties.push(
              t.objectProperty(
                t.identifier('classNamePrefix'),
                t.stringLiteral(createFileNameHash(filename)),
              ),
            );
          },
        });
      },
    },
  };
}

export default plugin;
