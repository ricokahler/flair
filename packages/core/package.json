{
  "name": "@react-style-system/core",
  "version": "0.0.0",
  "description": "",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "Rico Kahler",
  "license": "MIT",
  "peerDependencies": {
    "@types/react": "^16.8.0",
    "react": "^16.8.0"
  },
  "dependencies": {
    "@types/invariant": "^2.2.31",
    "invariant": "^2.2.4",
    "polished": "^3.4.4"
  }
}
