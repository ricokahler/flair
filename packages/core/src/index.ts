export { default as ColorContext } from './ColorContext';
export { default as ColorContextProvider } from './ColorContextProvider';
export { default as createReadablePalette } from './createReadablePalette';
export { default as ThemeContext } from './ThemeContext';
export { default as ThemeProvider } from './ThemeProvider';
export { default as useColorContext } from './useColorContext';
export { default as useTheme } from './useTheme';
export { default as css } from './css';
export { default as getContrast } from './getContrast';
export { default as getLuminance } from './getLuminance';
export { default as readableColor } from './readableColor';
export * from './types';
