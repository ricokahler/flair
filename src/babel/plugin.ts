import { Visitor } from '@babel/traverse';
import * as t from '@babel/types';
import createFileNameHash from 'src/common/createFileNameHash';

interface Result {
  visitor: Visitor;
}

function range(n: number) {
  return Array.from(Array(n)).map((_, i) => i);
}

function plugin(): Result {
  let foundCreateStyles = false;

  function createArrayPropertyValueFromTemplateLiteral(
    quasi: t.TemplateLiteral,
  ) {
    const { expressions, quasis } = quasi;

    const cssPropertyExpressions = range(expressions.length)
      .map(i => ({
        expression: expressions[i],
        templateElement: quasis[i],
      }))
      .filter(({ templateElement }) => {
        // must end with a `:` to signify that the expression is a CSS property
        return templateElement.value.raw.trim().endsWith(':');
      })
      .map(({ expression }) => expression);

    return t.arrayExpression(cssPropertyExpressions);
  }

  return {
    visitor: {
      ImportDeclaration(path, state: any) {
        const {
          importSourceValue = 'react-style-system',
          importedName = 'createStyles',
        } = state.opts;

        const { specifiers, source } = path.node;
        const hasCreateStyles = specifiers.some(node => {
          if (!t.isImportSpecifier(node)) return false;
          return node.imported.name === importedName;
        });
        if (!hasCreateStyles) return;

        const hasPackageName = source.value === importSourceValue;
        if (!hasPackageName) return;

        foundCreateStyles = true;
      },

      CallExpression(path, state: any) {
        const { opts, filename } = state;
        const { importedName = 'createStyles' } = opts;

        if (!foundCreateStyles) return;

        const { callee, arguments: expressionArguments } = path.node;
        if (!t.isIdentifier(callee)) return;
        if (callee.name !== importedName) return;

        const [firstArgument] = expressionArguments;

        if (
          !t.isFunctionExpression(firstArgument) &&
          !t.isArrowFunctionExpression(firstArgument)
        ) {
          return;
        }

        const { body } = firstArgument;

        let stylesObjectExpression!: t.ObjectExpression;
        if (t.isObjectExpression(body)) {
          stylesObjectExpression = body;
        } else {
          path.traverse({
            ReturnStatement(path) {
              const { argument } = path.node;

              if (!t.isObjectExpression(argument)) return;
              stylesObjectExpression = argument;
            },
          });
        }

        for (const property of stylesObjectExpression.properties) {
          if (!t.isObjectProperty(property)) return;
          const { value } = property;

          if (!t.isTaggedTemplateExpression(value)) return;
          const { tag, quasi } = value;

          if (!t.isIdentifier(tag)) return;
          if (tag.name !== 'css') return;

          const arrayExpression = createArrayPropertyValueFromTemplateLiteral(
            quasi,
          );

          property.value = arrayExpression;
        }

        stylesObjectExpression.properties.push(
          t.objectProperty(
            t.identifier('classNamePrefix'),
            t.stringLiteral(createFileNameHash(filename)),
          ),
        );
      },
    },
  };
}

export default plugin;
