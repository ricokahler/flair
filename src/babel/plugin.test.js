import { transform } from '@babel/core';
import fs from 'fs';
import path from 'path';
import plugin from './plugin';

it('removes the tagged template literals and replaces it with array expressions', async () => {
  const filename = path.resolve(__dirname, './example.js');
  const code = (await fs.promises.readFile(filename)).toString();

  const result = transform(code, {
    filename,
    presets: [],
    plugins: [plugin],
  });

  expect(result.code).toMatchInlineSnapshot(`
    "\\"use strict\\";

    var _reactStyleSystem = require(\\"react-style-system\\");

    var _polished = require(\\"polished\\");

    const useStyles = (0, _reactStyleSystem.createStyles)(({
      css,
      theme
    }) => ({
      root: [(0, _polished.readableColor)(theme.colors.brand), theme.colors.brand],
      title: [(0, _polished.darken)(0.1, theme.colors.brand)],
      classNamePrefix: \\"example--32e5a416\\"
    }));"
  `);
});
