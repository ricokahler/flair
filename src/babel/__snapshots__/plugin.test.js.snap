// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`removes the tagged template literals and replaces it with array expressions 1`] = `
"\\"use strict\\";

var _interopRequireDefault = require(\\"/Users/ricokahler/workspace/react-style-system/node_modules/@babel/runtime/helpers/interopRequireDefault\\");

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.default = void 0;

require(\\"/Users/ricokahler/workspace/react-style-system/src/examples/Example--d89a965.css\\");

var _react = _interopRequireDefault(require(\\"react\\"));

var _ssr = require(\\"react-style-system/ssr\\");

var _polished = require(\\"polished\\");

var _submodule = _interopRequireDefault(require(\\"./submodule\\"));

var _jsxFileName = \\"/Users/ricokahler/workspace/react-style-system/src/examples/Example.js\\";
const useStyles = (0, _ssr.createStyles)(({
  css,
  theme
}) => ({
  root: [(0, _polished.readableColor)(theme.colors.brand), theme.colors.brand],
  title: [(0, _polished.darken)(0.1, _submodule.default)],
  classNamePrefix: \\"Example--d89a965\\"
}));

function MyComponent(props) {
  const {
    Root,
    styles,
    title
  } = useStyles(props);
  return _react.default.createElement(Root, {
    __source: {
      fileName: _jsxFileName,
      lineNumber: 26
    },
    __self: this
  }, _react.default.createElement(\\"h1\\", {
    className: styles.title,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 27
    },
    __self: this
  }, title));
}

var _default = MyComponent;
exports.default = _default;"
`;
